lldb/source/Plugins/LanguageRuntime/Swift/SwiftMetadataCache.cpp

```diff
@@ -0,0 +1,294 @@
+#include "SwiftMetadataCache.h"
+
+#include "lldb/Utility/DataEncoder.h"
+#include "lldb/Utility/LLDBLog.h"
+#include "lldb/Utility/Log.h"
+#include "lldb/Version/Version.h"
+#include "llvm/CodeGen/AccelTable.h"
+#include "llvm/CodeGen/AsmPrinter.h"
+#include "llvm/Support/BLAKE3.h"
+#include "llvm/Support/CachePruning.h"
+#include "llvm/Support/Compression.h"
+
+using namespace lldb;
+using namespace lldb_private;
+using namespace swift::reflection;
+using namespace swift::remote;
+
+SwiftMetadataCache::SwiftMetadataCache() {
+  if (ModuleList::GetGlobalModuleListProperties()
+          .GetEnableSwiftMetadataCache()) {
+    llvm::CachePruningPolicy policy;
+    ModuleListProperties &properties =
+        ModuleList::GetGlobalModuleListProperties();
+    policy.Interval = std::chrono::hours(1);
+    policy.MaxSizeBytes = properties.GetSwiftMetadataCacheMaxByteSize();
+    policy.Expiration = std::chrono::hours(
+        properties.GetSwiftMetadataCacheExpirationDays() * 24);
+    m_data_file_cache.emplace(ModuleList::GetGlobalModuleListProperties()
+                                  .GetSwiftMetadataCachePath()
+                                  .GetPath(),
+                              policy);
+  }
+}
+
+SwiftMetadataCache *SwiftMetadataCache::get() {
+  if (!ModuleList::GetGlobalModuleListProperties()
+          .GetEnableSwiftMetadataCache()) 
+    return nullptr;
+
+  static SwiftMetadataCache instance;
+  return &instance;
+}
+
+void SwiftMetadataCache::registerModuleWithReflectionInfoID(ModuleSP module,
+                                                            uint64_t info_id) {
+  std::lock_guard<std::recursive_mutex> guard(m_mutex);
+  Log *log = GetLog(LLDBLog::Types);
+  auto it = m_info_to_module.find(info_id);
+
+  // We might've registered and loaded the cache for this module already. Return
+  // early in that case.
+  if (it != m_info_to_module.end()) {
+    if (std::get<bool>(it->second)) {
+      LLDB_LOGV(
+          log,
+          "[SwiftMetadataCache] Module {0} already register, returning early.",
+          module->GetFileSpec().GetFilename());
+      return;
+    }
+  }
+
+  m_info_to_module[info_id] = {module, false};
+  loadCacheForModule(module, info_id);
+}
+
+void SwiftMetadataCache::loadCacheForModule(ModuleSP module, uint64_t info_id) {
+  std::lock_guard<std::recursive_mutex> guard(m_mutex);
```

Comment:
Is this redundant with the lock in registerModuleWithReflectionInfoID or can this be called from other places? If not, should this be a lambda in registerModuleWithReflectionInfoID?

Reply:
It's redundant. My thinking was to guard this just in case we start calling it from somewhere else in the future, which would be a pretty subtle logic error. I'll merge the functions.